File: 2015/Day16/main.py
import re

def parse_input(filename):
    aunts = {}
    with open(filename, 'r') as file:
        for line in file:
            match = re.match(r'Sue (\d+): (.*)', line.strip())
            if match:
                sue_member = int(match.group(1))
                attributes = match.group(2).split(', ')
                aunt_data = {}
                for attr in attributes:
                    key, value = attr.split(': ')
                    aunt_data[key] = int(value)
                aunts[sue_member] = aunt_data
    return aunts

def find_matching_sue(aunts, mfcsam_reading):
    for sue, attributes in aunts.items():
        if all(mfcsam_reading[attr] == value for attr, value in attributes.items()):
            return sue
    return None

def match_attribute(attr, value, mfcsam_value):
    if attr in ['cats', 'trees']:
        return value > mfcsam_value if mfcsam_value is not None else True
    elif attr in ['pomeranians', 'goldfish']:
        return value < mfcsam_value if mfcsam_value is not None else True
    else:
        return value == mfcsam_value if mfcsam_value is not None else True

def find_real_sue(aunts, mfcsam_reading):
    for sue, attributes in aunts.items():
        if all(match_attribute(attr, value, mfcsam_reading.get(attr)) for attr, value in attributes.items()):
            return sue
    return None

mfcsam_reading = {
    'children': 3,
    'cats': 7,
    'samoyeds': 2,
    'pomeranians': 3,
    'akitas': 0,
    'vizslas': 0,
    'goldfish': 5,
    'trees': 3,
    'cars': 2,
    'perfumes': 1
}

aunts = parse_input('input.txt')

print(f"Part 1: {find_matching_sue(aunts, mfcsam_reading)}")
print(f"Part 2: {find_real_sue(aunts, mfcsam_reading)}")

File: 2015/Day16/main.dart
import 'dart:io';

void main() async {
  MfcsamReading mfcsamReading = {
    'children': 3,
    'cats': 7,
    'samoyeds': 2,
    'pomeranians': 3,
    'akitas': 0,
    'vizslas': 0,
    'goldfish': 5,
    'trees': 3,
    'cars': 2,
    'perfumes': 1
  };

  Aunts aunts = await parseInput('input.txt');

  print('Part 1: ${findMatchingSue(aunts, mfcsamReading)}');
  print('Part 2: ${findRealSue(aunts, mfcsamReading)}');
}

int? findMatchingSue(Aunts aunts, MfcsamReading mfcsamReading) {
  for (var entry in aunts.entries) {
    final sue = entry.key;
    final attributes = entry.value;
    if (attributes.entries
        .every((attr) => mfcsamReading[attr.key] == attr.value)) {
      return sue;
    }
  }
  return null;
}

int? findRealSue(Aunts aunts, MfcsamReading mfcsamReading) {
  for (var entry in aunts.entries) {
    final sue = entry.key;
    final attributes = entry.value;
    if (attributes.entries.every((attr) =>
        matchAttribute(attr.key, attr.value, mfcsamReading[attr.key]))) {
      return sue;
    }
  }
  return null;
}

bool matchAttribute(String attr, int value, int? mfcsamValue) {
  if (attr == 'cats' || attr == 'trees') {
    return mfcsamValue != null ? value > mfcsamValue : true;
  } else if (attr == 'pomeranians' || attr == 'goldfish') {
    return mfcsamValue != null ? value < mfcsamValue : true;
  } else {
    return mfcsamValue != null ? value == mfcsamValue : true;
  }
}

Future<Aunts> parseInput(String filename) async {
  Aunts aunts = {};
  final lines = await new File(filename).readAsLines();

  final regex = RegExp(r'Sue (\d+): (.*)');

  for (var line in lines) {
    final match = regex.firstMatch(line.trim());
    if (match != null) {
      final sueMember = int.parse(match.group(1)!);
      final attributes = match.group(2)!.split(', ');
      final aunt_data = <String, int>{};
      for (var attribute in attributes) {
        final parts = attribute.split(': ');
        aunt_data[parts[0]] = int.parse(parts[1]);
      }
      aunts[sueMember] = aunt_data;
    }
  }

  return aunts;
}

typedef Aunts = Map<int, Map<String, int>>;

typedef MfcsamReading = Map<String, int>;
